### ISSUE TITLE
Delegated generator return type is not respected (at random)

### ISSUE THREAD

**Issue by bagdonas** [microsoft/TypeScript#43721]
### 🔎 Search Terms
yield* delegated generator

### 🕗 Version & Regression Information
Since v3.6 (slightly different behaviour since v3.9)

### ⏯ Playground Link
[Playground link](https://www.typescriptlang.org/play?ts=4.3.0-beta#code/C4TwDgpgBAwghgGwQQQMbAJYHsB2UC8UA3qJAFxQDkqiClANFAGY4UAUAlAQHxQCuOANY4sAdxwBfALAAoACYRUCOACdoqXAGdgUGkgoAeAErc2Ldl3y8jl3gHEIOCCrjAsKg-CRpMuRkcYBYTEcbllZBSVVdS0dFQFDEzYAc0dnV3cLHigHJxc3Dy8UdGwcf0ChEXFuWygABRUsAFsMTQhjMJkIxWU1XVioVOAAMSwsLKsqJjHKbqi+jRxtQYhgACFVCd5KACNVWfke6P6lnSGjOBw5Zq2oHD4mnedwmXicMwES3AAqKE5iWRQE7LaZYAhQEAYCAIOS-PQIFKrUZYLhAtFQAD0GKgAANSBAsExmGMcVBWlBEKI4CBNLiAESguk4wHAuKXa5NcGQ6Gw3S0RHAC5XZpcLG4-GEqAuYVNUnkynU2k4+6PZzMmRAxbLPYqLlQmFw-lDDYqVHozHYvHgAlEnVy2kKmn0nVMllarAICAAOgQWGSZjGjGlHMYOo4sgk4a6rwEHxwXxwv3+RDdA1Bep5hqQAuRZvRYs0zVWGCaEE0vxxlxApLYmgAFlg+DCoE96YycQB+KOagbB5oZg187Pndkii0UhBUp3Kh5PFTqnunFuqAe8+ECk15tEFouYUvl3FVmv1xvN1s4ukuzvdqWrPgqPDpgDUy5UEajsjecYTSa4KY1rLEmChDcoO65DLm5pAjupZ7mWFZHn8J5NnILbQBe7ZdqmS59pyIH6muRqrEKHKitihawSW8GHjg1ZIQ2KFobiKpzte2Haiu+GZkOCLGqoW7QeRu5UQela0ceDFnuhl6qEyWEAWowD3o+YxQC+uFqa+74vF+TCfL4iZ-BwFC5OkBSeLQPilIw2gqBgODJBUwTVACAFajo6ZcWBREjGMAnjo6SoMmMrpub2o54RCBFZrxxERWRE5TkqLFquxOg6quMUbvxUEBZOirOrJC63kpD5AZpGkvjq2nRjBxb7r8paXJoshikCAC0aHTH0ABuADMXoAJxksAlAOvl05Hq12IdVA2iqJgDlQKIGDAHWUD9UNFJXPwYAjWNc3CfuNF0eyh2Ucda3QPiI1zZJqFuGh01op1V1QDdqB1pcqS0qIdaOOdMQKGStIQHIq1g38wAqCAFJyODS1OKIUAIPZZYUjob2OKhkpreSTAYJ6HBAA)

### 💻 Code
```ts
type CallAction = {type: 'call', fn: () => unknown}
declare const call: <R>(fn: () => R) => Generator<CallAction, R, unknown>

declare const run: <R>(generator: () => Generator<CallAction, R, unknown>) => Promise<R>

declare const getFoo: () => 'foo'
declare const getBar: () => 'bar'
declare const getRandom: () => number

run(function* () {
  const foo = yield* call(getFoo)       // `typeof foo` is always `"foo"`
  const random = yield* call(getRandom) // `typeof random` is always `number`
  const bar = yield* call(getBar)       // `typeof bar` is always `"bar"`
  console.log(foo, random, bar)
})

run(function* () {
  const foo = yield* call(getFoo)       // sometimes* `any` (should be `"foo"`?)
  const random = yield* call(getRandom) // always `number`
  const bar = yield* call(getBar)       // sometimes* `any` (should be `"bar"`?)
  return foo + bar
})

run(function* () {
  const foo = yield* call(getFoo)       // sometimes* `any` (should be `"foo"`?)
  const random = yield* call(getRandom) // sometimes* `any` (should be `number`?)
  const bar = yield* call(getBar)       // sometimes* `any` (should be `"bar"`?)
  return foo + random + bar
})

run(function* (): Generator<CallAction, string, unknown> {
  const foo = yield* call(getFoo)       // always `"foo"`
  const random = yield* call(getRandom) // always `number`
  const bar = yield* call(getBar)       // always `"bar"`
  return foo + random + bar
})

// sometimes* means
//   - before v3.9 it's always `any`
//   - starting with v3.9 and up it's sometimes `any` and sometimes the type it should to be
//   - the type changes when some code is edited (try adding new lines at the end of this file)
```

### 🙁 Actual behavior
The type of value of `yield* {generator}` is randomly `any` if outer generator function doesn't have its return type explicitly declared. 

### 🙂 Expected behavior
The type of value of `yield* {generator}` should not be randomly `any`

**Comment by bagdonas** [comment]
After some further digging it looks to me like this is probably a duplicate of #39812. If that's the case, apologies for posting this issue.

**Comment by rbuckton** [comment]
This seems like it may have been addressed in TypeScript 4.7 as it no longer repros.
